src/Microsoft.VisualStudio.Threading.Shared/AsyncReaderWriterLock.cs

```diff
@@ -1837,16 +1837,19 @@ public void Dispose()
                         using (nonConcurrentSyncContext != null ? nonConcurrentSyncContext.LoanBackAnyHeldResource(this.awaiter.OwningLock) : default(NonConcurrentSynchronizationContext.LoanBack))
                         {
                             var releaseTask = this.awaiter.ReleaseAsync();
-                            try
+                            using (NoMessagePumpSyncContext.Default.Apply())
```

Comment:
Most of our internal locks (and all of JTF) do this. We missed this spot. The way we discovered it is VS crashed (or hung? I can't remember) due to the reentrancy invited by this unprotected wait while on the UI thread.